module/Database/src/Mapper/ProspectiveMember.php

```diff
@@ -125,38 +126,53 @@ public function find(int $lidnr): ?ProspectiveMemberModel
      * @return ProspectiveMemberModel[]
      */
     public function findWithFullyExpiredOrFailedCheckout(): array
     {
+        // Get all prospective members and their checkout sessions
         $qb = $this->getRepository()->createQueryBuilder('m');
         $qb->leftJoin(CheckoutSessionModel::class, 'cs', Join::WITH, 'cs.prospectiveMember = m.lidnr');
 
+        // Subquery to get maximum checkout session for a member
         $qbc = $this->em->createQueryBuilder();
-        $qbc->select('MAX(css.id)')
+        $qbc->select('MAX(csm.id)')
+            ->from(CheckoutSessionModel::class, 'csm')
+            ->where('csm.prospectiveMember = m.lidnr');
+
+        // Subquery to get the original (expired) checkout session (the one that could be recovered)
+        $qbd = $this->em->createQueryBuilder();
+        $qbd->select('(CASE WHEN css.recoveredFrom IS NOT NULL THEN IDENTITY(css.recoveredFrom) ELSE css.id END)')
             ->from(CheckoutSessionModel::class, 'css')
-            ->where('css.prospectiveMember = m.lidnr');
-        $qb->where($qb->expr()->eq('cs.id', '(' . $qbc->getDQL() . ')'))
-            ->andWhere($qb->expr()->orX(
-                $qb->expr()->andX(
-                    $qb->expr()->eq('cs.state', ':expired'),
-                    $qb->expr()->lt('cs.expiration', ':fullyExpired'),
-                ),
-                $qb->expr()->andX(
-                    $qb->expr()->eq('cs.state', ':failed'),
-                    $qb->expr()->lt('cs.expiration', ':fullyFailed'),
-                ),
-            ));
+            ->where('css.prospectiveMember = m.lidnr')
+            ->andWhere($qb->expr()->eq('css.id', '(' . str_replace('csm', 'csm2', $qbc->getDQL()) . ')'))
+            ->andWhere('css.state = :expired');
+
+        $qb->where($qb->expr()->orX(
+            // Get the last checkout session, if it has failed more than 31 days ago
+            $qb->expr()->andX(
+                $qb->expr()->eq('cs.id', '(' . $qbc->getDQL() . ')'),
+                $qb->expr()->eq('cs.state', ':failed'),
+                $qb->expr()->lt('cs.expiration', ':fullyFailed'),
+            ),
+            // OR get the original session if it has expired more than a day ago using that
+            // if x.state == Expired, the expiration date is the last date the checkout session can be recoverd
+            $qb->expr()->andX(
+                $qb->expr()->eq('cs.id', '(' . $qbd->getDQL() . ')'),
+                $qb->expr()->eq('cs.state', ':expired'),
+                $qb->expr()->lt('cs.expiration', ':fullyExpired'),
+            ),
+        ));
 
         $qb->setParameter('expired', CheckoutSessionStates::Expired)
             ->setParameter('failed', CheckoutSessionStates::Failed)
             ->setParameter('fullyExpired', (new DateTime())->sub(new DateInterval('P1D')))
             ->setParameter('fullyFailed', (new DateTime())->sub(new DateInterval('P31D')));
 
-        return $qb->getQuery()->getResult();
+        return [];
```

Comment:
@rinkp can you put the `$qb->getQuery()->getResult()` back?

Reply:
:-( 
I was so sure I had removed all debugging statements

[aac78f261c25802009e30116653b1739940a96c7..897407dccb29a75656f7ab800fb443d38604a63f](https://github.com/GEWIS/gewisdb/compare/aac78f261c25802009e30116653b1739940a96c7..897407dccb29a75656f7ab800fb443d38604a63f)

Reply:
Thanks!